processing/src/main/java/org/apache/druid/query/scan/ScanQueryEngine.java

```diff
@@ -252,7 +260,295 @@ public void cleanup(Iterator<ScanResultValue> iterFromMake)
             ))
     );
   }
+  private Sequence<ScanResultValue> getScanOrderByResultValueSequence(
```

Comment:
Is the scan query engine the right place for a sort? Scan queries are distributed. Each data node scans a separate segment. Should we do a sort per data node? I suppose yes, since then the Broker need only merge.

The Druid pattern seems to be to insert another `QueryRunner`/`Sequence` pair for a new task. So, we'd want to insert a "sorting query runner" somewhere so that the sorting operation can be separate from the "segment scan" operation. This split avoids making the scan query engine more complex than it already is.

Also, where should the sort occur? Per cursor? Per segment? Per node? If per node, then only the Broker merges. If per node (i.e., if node 1 scans 5 segments, sort the results of all 5), then the operation has to be done above the scan query engine level: at the point where we have visibility to all the data for the node.

If per segment, then we have to merge per-node and in the Broker.

I wonder, does this draft handle these cases?

Reply:
> 


This is a good idea. Query engines should not be responsible for sorting. Let me see how to insert a "Sorting Query runner".

I think that currently, orderby is implemented in a violent way. If we follow the  Druid pattern , the insertion and sorting position is not suitable.

The current processing method is as follows:

Segments are sorted in `ScanqueryEngine.process`.

Data nodes are sorted in `ScanqueryRunnerFactory.mergeRunners`.

The broker is sorted in `ScanqueryQueryToolCheck.mergeReults`

Reply:
> 

Today I thought about why I changed it to this way in the first place.
Take a look, in fact `orderByTimeLimit` and `orderByMultiColumnLimit` are similar.
```
//past execution path
if(orderByTimeLimit){
  //...
}else{
  //...
}

//current execution path
if(orderByMultiColumnLimit){
  //
}else if(orderByTimeLimit){
  ///
}else{
  ///
}
```